### PR TITLE
[#PLAT-211]Determine if registration can be done entirely via API

### PR DESCRIPTION
**Pull request by chennan47**
screen shot 2018-02-02 at 2 59 41 pm

screen shot 2018-02-02 at 2 59 49 pm

screen shot 2018-02-02 at 3 00 07 pm

https://openscience.atlassian.net/browse/PLAT-211

### CODE REVIEW

File: swagger-spec/registrations/definition.yaml [hunk 1]
```diff
@@ -141,6 +144,19 @@ properties:
...
         readOnly: true
         description: 'The reasoning for why this registration was withdrawn, as entered by the administrator that initiated the withdrawal.'
+      draft_registration:
+        type: string
+        readOnly: false
+        description: 'The unique ID of the draft registration from which the registration was created.'
+      registration_choice:
+        type: string
+        readOnly: false
+        description: 'Type of registration you want to choose to be, immediate or embargo. If you choose to be embargo, you need to provide with a lift_embargo datetime.'
```
Comment (pattisdr) [inline swagger-spec/registrations/definition.yaml]: Rewording: just a suggestion:
Describes when the registration will become public, either "immediate" or "embargo".  If this field is "embargo", you will need to supply a datetime in the "lift_embargo" field.

File: swagger-spec/registrations/definition.yaml [hunk 2]
```diff
@@ -141,6 +144,19 @@ properties:
         type: string
         readOnly: true
         description: 'The reasoning for why this registration was withdrawn, as entered by the administrator that initiated the withdrawal.'
+      draft_registration:
+        type: string
+        readOnly: false
+        description: 'The unique ID of the draft registration from which the registration was created.'
```
Comment (pattisdr) [inline swagger-spec/registrations/definition.yaml]: Rewording suggestion: The ID of the draft registration from which the registration will be created.

File: swagger-spec/registrations/definition.yaml [hunk 3]
```diff
@@ -141,6 +144,19 @@ properties:
...
+      registration_choice:
+        type: string
+        readOnly: false
+        description: 'Type of registration you want to choose to be, immediate or embargo. If you choose to be embargo, you need to provide with a lift_embargo datetime.'
+      lift_embargo:
+        type: string
+        format: date-time
+        readOnly: false
+        description: 'The time at which the registered node will become public only avaible when you choose embargo for registration_choice.'
```
Comment (pattisdr) [inline swagger-spec/registrations/definition.yaml]: Rewording suggestion: A future datetime when the registration will become public.  This field should be set when "registration_choice" is set to "embargo".

File: swagger-spec/nodes/registrations_list.yaml [hunk 4]
```diff
@@ -182,3 +182,64 @@ get:
...
+    #### Permissions
+
+    Only project administrators may view create registrations.
+
+    #### Required
+
+    Required fields for creating a registration include:
+
+
+    &nbsp;&nbsp;&nbsp;&nbsp`type`
```
Comment (pattisdr) [inline swagger-spec/nodes/registrations_list.yaml]: Any way to tab this without typing &nbsp...? Locally, this still looks like &nbsp.

File: swagger-spec/nodes/registrations_list.yaml [hunk 5]
```diff
@@ -182,3 +182,64 @@ get:
               node_license:
             type: registrations
             id: esa63
+
+# /nodes/{node_id}/registrations/
+post:
+  summary: Create a registration
+  description: >-
+    Initiate a registration of the current draft registration.
```
Comment (pattisdr) [inline swagger-spec/nodes/registrations_list.yaml]: I don't know if I'd use the word "initiate" here.  This is really the final step in a multi-step process of creating a registration.  "Finalize the registration process." , maybe, followed by:

A registration is created from a completed draft registration (containing the supplemental registration questions), information about whether the registration will become public immediately or be embargoed, and conditionally, the datetime at which the registration will become public.  A registration is a frozen version of the project that can never be edited or deleted, but can be withdrawn. Your original project remains editable but will now have the registration linked to it.

File: swagger-spec/nodes/registrations_list.yaml [hunk 6]
```diff
@@ -182,3 +182,64 @@ get:
...
+
+
+    &nbsp;&nbsp;&nbsp;&nbsp`type`
+
+
+    &nbsp;&nbsp;&nbsp;&nbsp`draft_registration`
+
+
+    &nbsp;&nbsp;&nbsp;&nbsp`registration_choice`
+
```
Comment (pattisdr) [inline swagger-spec/nodes/registrations_list.yaml]: I would add `lift_embargo` here and add a note that it is conditionally required on whether registration_choice is `embargo`

File: swagger-spec/nodes/draft_registration_detail.yaml [hunk 7]
```diff
@@ -83,7 +83,13 @@ patch:
   summary: Update a draft registration
   description: >-
     Updates a draft registration by setting the values of the attributes specified in the request body.
-    Any unspecified attributes will be left unchanged.
+    Any unspecified attributes will be left unchanged. It is the second step in creating a registration.
+
+
+    Refer to step 1: [How to create a draft registration](#operation/nodes_draft_registrations_create)
```
Comment (pattisdr) [inline swagger-spec/nodes/draft_registration_detail.yaml]: Thanks for all these links, this is perfect 👍

File: swagger-spec/nodes/draft_registration_detail.yaml [hunk 8]
```diff
@@ -83,7 +83,13 @@ patch:
   summary: Update a draft registration
   description: >-
     Updates a draft registration by setting the values of the attributes specified in the request body.
-    Any unspecified attributes will be left unchanged.
```
[the remark this answers is not in the document]
Comment (pattisdr) [reply swagger-spec/nodes/draft_registration_detail.yaml]: Then delete L95-107
> Draft registrations contain the supplemental registration questions that accompany a registration.
     Answer the questions in the draft registration supplement by sending update requests until you are ready to submit the draft. The registration supplement of a draft registration cannot be updated after the draft has been created. When updating a draft registration, `registration_metadata` is required. It must be a dictionary with keys as question ids in the registration form, and values as nested dictionaries matching the specific format in the [registration schema](TODO: link me pls).  If a question is multiple-choice, the question response must exactly match one of the possible choices.

File: swagger-spec/nodes/draft_registration_detail_definition.yaml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    readOnly: true
+    required:
+      - html
+    description: 'URLs to alternative representations of the draft registration entity.'
+    properties:
+      html:
+        type: string
+        format: URL
+        readOnly: true
+        description: 'A link to the draft registration''s page on the OSF.'
```
Comment (pattisdr) [inline swagger-spec/nodes/draft_registration_detail_definition.yaml]: Can you tell me where to find this section in the docs? The draft registration detail definition? Also, there's a double `''`.

File: swagger-spec/nodes/registrations_list.yaml [hunk 10]
```diff
@@ -182,3 +182,74 @@ get:
               node_license:
             type: registrations
             id: esa63
+
+# /nodes/{node_id}/registrations/
```
[the remark this answers is not in the document]
Comment (chennan47) [reply swagger-spec/nodes/registrations_list.yaml]: Tickets made: https://openscience.atlassian.net/browse/PLAT-660

**Comment by sloria** [comment]
Awesome 🚢

**Pull request merged**